### PR TITLE
8263352: assert(use == polladr) failed: the use should be a safepoint polling

### PR DESCRIPTION
**Pull request by Wanghuang-Huawei**
The main reason of this failure has been shown in JDK-8263352.
* JDK-8260637 makes a CastPP unpinned, so it can be pull out of inner loop
* JDK-8247307 assert use == polladr and makes it crash here.
* I think we can only `intrinsify_fill` the array when it is *really* safe here, so I fix this bug by `return` other cases.
In this patch, I will give a small test case called `Test8263352.java` to reproduct this failure.
Thanks to @vnkozlov , who found this failure for the first time (to me).
Thanks to  @nsjian，who found the case which can reproduct this failure everytime.
Thanks to @pfustc, who changed notes with us.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263352](https://bugs.openjdk.java.net/browse/JDK-8263352): assert(use == polladr) failed: the use should be a safepoint polling

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 7acd73d561af4d6095e4104ea6531c2df1037de7

### Contributors
 * Wang Huang `<whuang@openjdk.org>`
 * Wu Yan `<wuyan34@huawei.com>`

### Download
To checkout this PR locally:
`$ git fetch https://git.openjdk.java.net/jdk pull/3061/head:pull/3061`
`$ git checkout pull/3061`

To update a local copy of the PR:
`$ git checkout pull/3061`
`$ git pull https://git.openjdk.java.net/jdk pull/3061/head`

### CODE REVIEW

**Comment by vnkozlov** [comment]
I have question for @rwestrel who did strip mining. Looking on nodes I see:
```
 1247  CountedLoop  ===  1247  1246  794  [[ 1247  788  741  738  781  1242 ]] inner stride: 1  strip mined !orig=[1237],[731] !jvms: ComponentSampleModel::<init> @ bci:147 (line 162)
 1242  CountedLoopEnd  ===  1247  1241  [[ 1243  794 ]] [lt] P=0.646827, C=6564.000000 !orig=[793] !jvms: ComponentSampleModel::<init> @ bci:144 (line 161)
 1243  IfFalse  ===  1242  [[ 1248 ]] #0 !orig=795 !jvms: ComponentSampleModel::<init> @ bci:144 (line 161)

 788  LoadP  ===  1247  659  787  [[ 1248 ]]  @rawptr:BotPTR, idx=Raw; #rawptr:BotPTR (does not depend only on test) !jvms: ComponentSampleModel::<init> @ bci:158 (line 161)
 1248  SafePoint  ===  1243  1  784  1  1  788  10  1  1  1  1  1  1  782  [[ 1244 ]]  SafePoint  !orig=783 !jvms: ComponentSampleModel::<init> @ bci:158 (line 161)
```
So why Poll's LoadP node's control edge  is not the same as SafePoint's control edge? Why it points to inner loop head?
If we adjust LoadP's control edge to the same node as Safepoint (exit from inner loop), we would not need this special case in match_fill_loop().

File: test/hotspot/jtreg/compiler/c2/Test8263352.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
+/*
```
Comment (vnkozlov) [inline test/hotspot/jtreg/compiler/c2/Test8263352.java:1]: We don't use bug ID for tests anymore. Rename it to something meaningful and move it to corresponding `compiler/loopopts/` or `compiler/loopstripmining/` directory.

**Comment by rwestrel** [comment]
> I have question for @rwestrel who did strip mining. Looking on nodes I see:
> 
> ```
>  1247  CountedLoop  ===  1247  1246  794  [[ 1247  788  741  738  781  1242 ]] inner stride: 1  strip mined !orig=[1237],[731] !jvms: ComponentSampleModel::<init> @ bci:147 (line 162)
>  1242  CountedLoopEnd  ===  1247  1241  [[ 1243  794 ]] [lt] P=0.646827, C=6564.000000 !orig=[793] !jvms: ComponentSampleModel::<init> @ bci:144 (line 161)
>  1243  IfFalse  ===  1242  [[ 1248 ]] #0 !orig=795 !jvms: ComponentSampleModel::<init> @ bci:144 (line 161)
> 
>  788  LoadP  ===  1247  659  787  [[ 1248 ]]  @rawptr:BotPTR, idx=Raw; #rawptr:BotPTR (does not depend only on test) !jvms: ComponentSampleModel::<init> @ bci:158 (line 161)
>  1248  SafePoint  ===  1243  1  784  1  1  788  10  1  1  1  1  1  1  782  [[ 1244 ]]  SafePoint  !orig=783 !jvms: ComponentSampleModel::<init> @ bci:158 (line 161)
> ```
> 
> So why Poll's LoadP node's control edge is not the same as SafePoint's control edge? Why it points to inner loop head?
> If we adjust LoadP's control edge to the same node as Safepoint (exit from inner loop), we would not need this special case in match_fill_loop().

That makes sense but needs to be done explicitly:
```
diff --git a/src/hotspot/share/opto/loopnode.cpp b/src/hotspot/share/opto/loopnode.cpp
index bf5a685960c..42e5b940dc5 100644
--- a/src/hotspot/share/opto/loopnode.cpp
+++ b/src/hotspot/share/opto/loopnode.cpp
@@ -1636,6 +1636,13 @@ bool PhaseIdealLoop::is_counted_loop(Node* x, IdealLoopTree*&loop, BasicType iv_
         Node* sfpt = sfpt2->clone();
         sfpt->set_req(0, iffalse);
         outer_le->set_req(0, sfpt);
+        Node* polladdr = sfpt->in(TypeFunc::Parms)->clone();
+        assert(polladdr->is_Load(), "");
+        Node* new_polladdr = polladdr->clone();
+        new_polladdr->set_req(0, iffalse);
+        _igvn.register_new_node_with_optimizer(new_polladdr, polladdr);
+        set_ctrl(new_polladdr, iffalse);
+        sfpt->set_req(TypeFunc::Parms, new_polladdr);
         // When this code runs, loop bodies have not yet been populated.
         const bool body_populated = false;
         register_control(sfpt, outer_ilt, iffalse, body_populated);
```

**Comment by vnkozlov** [comment]
Yes, that is it.

@rwestrel  why you need first clone?
```
Node* polladdr = sfpt->in(TypeFunc::Parms)->clone();
```

I think we can use original one because it is only used to copy notes in  `register_new_node_with_optimizer()`

And do we always have this Load? May be runtime check instead of assert?

**Comment by rwestrel** [comment]
> @rwestrel why you need first clone?

Right. It's likely not needed;

> And do we always have this Load? May be runtime check instead of assert?

That I don't know.

**Comment by vnkozlov** [comment]
@Wanghuang-Huawei are you working on Roland's implementation? Or you want us to work on it?

**Review by vnkozlov — APPROVED** [review]
Okay. Lets push the fix as suggested in match_fill_loop() and new test.
And I will file separate RFE for Roland's changes.

**Comment by vnkozlov** [comment]
I filed https://bugs.openjdk.java.net/browse/JDK-8264063

**Comment by Wanghuang-Huawei** [comment]
> Okay. Lets push the fix as suggested in match_fill_loop() and new test.
> And I will file separate RFE for Roland's changes.

Thank you for your review. I will rename the test case file so that it can be used as test case of 

> @Wanghuang-Huawei are you working on Roland's implementation? Or you want us to work on it?

I am not sure if  `polladdr->is_Load()` is always true? 

> Okay. Lets push the fix as suggested in match_fill_loop() and new test.
> And I will file separate RFE for Roland's changes.

Thank you for your review. I've renamed the test case (which can also be used as test case of JDK-8264063 ?) and pushed that.

**Comment by Wanghuang-Huawei** [comment]
> I filed https://bugs.openjdk.java.net/browse/JDK-8264063

Thank you for your work !

**Comment by vnkozlov** [comment]
Yes, the test will be very useful for 8264063.

**Pull request closed without merging**